dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/importer/context/ProgramEventSupplier.java

```diff
@@ -0,0 +1,424 @@
+                        program.setTrackedEntityType( trackedEntityType );
+                    }
+
+                    program.setProgramStages( programStages );
+                    results.put( getProgramKey( idScheme, rs ), program );
+
+                    programId = program.getId();
+                }
+                else
+                {
+                    results.get( getProgramKey( idScheme, rs ) ).getProgramStages()
+                        .add( toProgramStage( rs ) );
+                }
+            }
+            return results;
+        } ) ).orElse( new HashMap<>() );
+    }
+
+    /**
+     * Resolve the key to place in the Program Map, based on the Scheme
+     * specified in the request If the scheme is of type Attribute, use the
+     * attribute value from the JSONB column
+     */
+    private String getProgramKey( IdScheme programIdScheme, ResultSet rs )
+        throws SQLException
+    {
+        return programIdScheme.isAttribute() ? rs.getString( ATTRIBUTESCHEME_COL )
+            : IdSchemeUtils.getKey( programIdScheme, rs );
+    }
+
+    private ProgramStage toProgramStage( ResultSet rs )
+        throws SQLException
+    {
+        ProgramStage programStage = new ProgramStage();
+        programStage.setId( rs.getLong( "ps_id" ) );
+        programStage.setUid( rs.getString( "ps_uid" ) );
+        programStage.setCode( rs.getString( "ps_code" ) );
+        programStage.setName( rs.getString( "ps_name" ) );
+        programStage.setSortOrder( rs.getInt( "sort_order" ) );
+        programStage.setSharing( jsonToEntity( rs.getString( "ps_sharing" ), Sharing.class ) );
+        programStage.setFeatureType(
+            rs.getString( "ps_feature_type" ) != null ? FeatureType.valueOf( rs.getString( "ps_feature_type" ) )
+                : FeatureType.NONE );
+        programStage.setRepeatable( rs.getBoolean( "ps_repeatable" ) );
+        programStage.setEnableUserAssignment( rs.getBoolean( "enableuserassignment" ) );
+        String validationStrategy = rs.getString( "validationstrategy" );
+        if ( StringUtils.isNotEmpty( validationStrategy ) )
+        {
+            programStage.setValidationStrategy( ValidationStrategy.valueOf( validationStrategy ) );
+        }
+        return programStage;
+    }
+
+    private DataElement toDataElement( ResultSet rs )
+        throws SQLException
+    {
+        DataElement dataElement = new DataElement();
+        dataElement.setId( rs.getLong( "dataelementid" ) );
+        dataElement.setUid( rs.getString( "de_uid" ) );
+        dataElement.setCode( rs.getString( "de_code" ) );
+        return dataElement;
+    }
+
+    /**
+     * Check if the list of incoming Events contains one or more Program uid
+     * which is not in cache. Reload the entire program cache if a Program UID
+     * is not found
+     */
+    private boolean requiresCacheReload( List<Event> events, Map<String, Program> programMap )
```

Comment:
Maybe not part of this PR but at some point it will be good to do something smarter here and reload the cache less frequently.